### ISSUE TITLE
setVisible(false) on PolymerTemplate does not work if the PolymerTemplate is within a component with a shadow root.

### ISSUE THREAD

**Issue by Peppe** [vaadin/flow#8256]
### Description of the bug
When building apps using `PolymerTemplate`, `polymerTemplate.setVisible(false)` does not work. Normally `setVisible(false)` applies the `hidden="true"` attribute, and Vaadin has the generic CSS rule available `:host([hidden]) { display: none !important; }` which hides the component. Things break if the component you try to hide is within another component that uses a shadow root. The attribute is added correctly, but the CSS is not applied. If you have for example a VerticalLayout within a component that has a shadow root, then setVisible() works correctly, so the bug has so far only appeared with PolymerTemplate

### Expected behavior
VisibilityBug.java
```java
package com.vaadin.tutorial.crm.ui;

import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.Paragraph;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.router.Route;

@Route("visibility-bug")
public class VisibilityBug extends VerticalLayout {

    boolean isVisible = true;

    public VisibilityBug() {
        VerticalLayout verticalLayout1 = new VerticalLayout(new Span("Hello VerticalLayout"));
        MyTemplateOuter template1 = new MyTemplateOuter();

        Button button1 = new Button("Toggle visibility when no shadow doms are present", e -> {
            isVisible = !isVisible;
            verticalLayout1.setVisible(isVisible);
            template1.setVisible(isVisible);
        });
        add(button1, verticalLayout1, template1);

        add(new Paragraph());

        MyTemplateOuter template2 = new MyTemplateOuter();

        Button button2 = new Button("Toggle visibility when a shadow dom is present", e -> {
            isVisible = !isVisible;
            template2.setInnerVisibilities(isVisible);
        });
        add(button2, template2);
    }
}
```
MyTempalteOuter.java
```java
package com.vaadin.tutorial.crm.ui;

import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.polymertemplate.Id;
import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.templatemodel.TemplateModel;

@Tag("my-template-outer")
@JsModule("./src/views/my-template-outer.js")
public class MyTemplateOuter extends PolymerTemplate<MyTemplateOuter.MyTemplateOuterModel> {

    @Id("vl") private VerticalLayout vl;

    @Id("inner") private MyTemplateInner innerTemplate;

    public MyTemplateOuter() { }
    public interface MyTemplateOuterModel extends TemplateModel {}
    public void setInnerVisibilities(boolean visible){
        vl.setVisible(visible);
        innerTemplate.setVisible(visible);
    }
}
```
MyTemplateInner.java
```java
package com.vaadin.tutorial.crm.ui;

import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.polymertemplate.PolymerTemplate;
import com.vaadin.flow.templatemodel.TemplateModel;

@Tag("my-template-inner")
@JsModule("./src/views/my-template-inner.js")
public class MyTemplateInner extends PolymerTemplate<MyTemplateInner.MyTemplateInnerModel> {
    public MyTemplateInner() {}
    public interface MyTemplateInnerModel extends TemplateModel {}
}
```

my-template-outer.js
```js
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
import './my-template-inner.js';
import '@vaadin/vaadin-ordered-layout/src/vaadin-vertical-layout.js';

class MyTemplateOuter extends PolymerElement {

    static get template() {
        return html`
<style include="shared-styles">
                :host {
                    display: block;
                    height: 100%;
                    border: 1px solid black;
                    padding: 10px;
                }
            </style>
            <vaadin-vertical-layout id="vl">
              <span>Hello inner VerticalLayout</span>
            </vaadin-vertical-layout>
            <my-template-inner id="inner"></my-template-inner>
`;
    }

    static get is() {
        return 'my-template-outer';
    }

    static get properties() {
        return {
            // Declare your properties here.
        };
    }
}
customElements.define(MyTemplateOuter.is, MyTemplateOuter);
```

my-template-inner.js
```js
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';

class MyTemplateInner extends PolymerElement {

    static get template() {
        return html`
<style include="shared-styles">
                :host {
                    display: block;
                    height: 100%;
                }
            </style>
<div>Hello template inner</div>
`;
    }

    static get is() {
        return 'my-template-inner';
    }

    static get properties() {
        return {
            // Declare your properties here.
        };
    }
}

customElements.define(MyTemplateInner.is, MyTemplateInner);
```

### Actual behavior
A PolymerTemplate stays visible after `setVisible(false)` if that PolymerTemplate is within a shadow root. `hidden="true"` is applied but not the CSS

![vaadin-template-visibility](https://user-images.githubusercontent.com/525503/81157134-578a5400-8f8f-11ea-87b5-e49e98487d41.gif)

### Expected behavior
A component that you call `setVisible(false)` on should be hidden

### Versions:
    - Vaadin / Flow version: 14.1.21
    - Java version: Corretto 11
    - OS version: Windows 10

**Comment by Peppe** [comment]
A workaround to this problem is now part of the Designer tutorial: https://vaadin.com/learn/tutorials/vaadin-designer-tutorial#java-hiding-the-form

**Comment by mvysny** [comment]
Another workaround could be:

```
getElement().executeJs("this.style.display='none'");
```

**Comment by denis-anisimov** [comment]
Thanks a lot @Peppe for exact, detailed and short steps to reproduce!

`setVisible()` works as designed: it adds `hidden` attribute to the element.
Apparently it's not enough in some cases (the example is above).
There a couple of issues:
- it's not clear which approach should be used to use on the client side instead of `hidden`.
- it may quite dangerous change : it may introduce regressions.

Considering two issues together I cannot proceed with this without exact design.
It has to be discussed and design decision is required.

@Legioth, could you please share you opinion about this?

**Comment by Legioth** [comment]
The problem here is indeed that this case leads to a situation where there isn't any CSS selector that applies `display: none` based on the `hidden` attribute. It happens to work with most Vaadin components because they have their own `:host([hidden])` selector, but this isn't automatically the case for arbitrary HTML elements.

I'm currently only aware of alternatives that are all problematic for various reasons.
1. Change Flow to apply `display: none` as an inline style instead of or in addition to the `hidden` attribute.
2. Change Flow to automatically inject the appropriate CSS inside the shadow root of PolymerTemplate instances.
3. Change Designer to automatically include the appropriate CSS in all template files that it creates.
4. wontfix

The best alternative would of course be that someone could come up with some alternative solution that wouldn't be as problematic. Otherwise, we'd have to make a tradeoff to determine which alternative is least bad.

**Comment by StefanCov** [comment]
So I should use the workarounds for this issue? This is always a problem within my application as MainView is a PolymerTemplate along with my custom components. I'm unable to set the visibility on any of these and have to jump through loops to hide them from the user.

**Comment by denis-anisimov** [comment]
We are discussing this now: what should we do with this.
For the current projects you can use a workaround.

**Comment by StefanCov** [comment]
Since I'll be using the workaround for now should I remove my warranty and let yous solve this at your own rate? I have no rush for this if using the workout is good enough for now.

**Comment by denis-anisimov** [comment]
I cannot answer to this question.
One of the discussed options is `wontfix`.
So it's theoretically possible that we won't do anything with this and let users apply a workaround (e.g. with `:host([hidden])` selector) on their side.

**Comment by Legioth** [comment]
> This is always a problem within my application as MainView is a PolymerTemplate along with my custom components

My recommended workaround for those situations would be to include `[hidden] { display: none; }` as a style in each template. You can create your own subclass (either server-side for `PolymerTemplate` or client-side for `PolymerElement`) that would automatically apply the style and use that one consistently throughout the application so that you wouldn't have to separately include the extra style separately in each template.

You can indeed remove the prioritization if you conclude that the workaround is good enough for you.

**Comment by StefanCov** [comment]
Yeah, I will go with the workaround. Thanks!

The workaround is good enough. I set to priority just to confirm if this was indeed a bug that would be fixed. Just making sure I wouldn't be wasting my time with workarounds if it would get fixed by yourselves but the answer is clear now.

**Comment by Legioth** [comment]
If this issue is prioritized, then we are "forced" to resolve it in one way or another quite soon. If the issue is not prioritized, then we may choose put our focus on some other issue instead (or expedite work on other prioritized issues).

**Comment by denis-anisimov** [comment]
@Legioth , it's a BFP.
It's already prioritized.

**Comment by StefanCov** [comment]
I've removed the warranty, probably best spending time on an issue that doesn't have a workaround.

**Comment by Peppe** [comment]
Thank you for looking into the matter and showing that you have put thought into this issue. 

I still very much would like to see a fix for this. It is kinda bad that you lose the capability of using `setVisible()` without workarounding it in your project, if you've built your UI's with PolymerTemplate. 

In the [Designer tutorial, we have now a section](https://vaadin.com/learn/tutorials/vaadin-designer-tutorial#java-hiding-the-form) warning people on this bug and instructing them to put a workaround in place. It doesn't give much confidence for tutorial-readers to need to resort to workarounds in basic cases.

**Comment by Artur-** [comment]
Why is it "problematic" so set `display: none` for `PolymerTemplate` instead of a `hidden` class that has no effect?

**Comment by Legioth** [comment]
CSS selectors that target the `hidden` attribute can be overridden by something with higher specificity, whereas an inline style would require an `!important` style. It's quite a special case that anyone has overridden `hidden` styles to keep some parts visible, but I've heard about weirder workarounds.

It would thus be better if some less obtrusive workaround could be found, but it doesn't seem like anybody knows what that would be.

**Comment by denis-anisimov** [comment]
No problems at all.
It's a source of regressions.

**Comment by Peppe** [comment]
Is `hidden` the only case where Lumo/Vaadin adds css? I wonder if fixing `hidden` is just one fix of many that might be needed and thus if 2. would be a better option?

**Comment by Artur-** [comment]
Being able to override server side visibility with CSS sounds more like a bug than a feature to me.

I don't think Flow should modify the template in any way, that would be magic that nobody can understand.

**Comment by StefanCov** [comment]
Yeah, the issue I was having was when I was using `getElement().getStyle().set("display","none")` it would not work or get overridden. It also took me some time to realise I could not get `setVisible()` working either. Which was extremely hard to debug as I had no real idea why it would not work when it works in other places.

My code just ended up very inconsistent using multiple different methods to hide specific parts of my application. 

The weird part was that `getElement().executeJs("this.style.display='none'")` would work but `getElement().getStyle().set("display","none")` didn't, that might just be a different issue.

**Comment by pleku** [comment]
Now that we're looking into improving the DX for "building layouts with Designer and having UI logic in Java", I think this issue should be fixed to be able to have _good enough DX_. I agree with @Peppe that if `myTemplateComponent.setVisible(false);` doesn't work, it is not a good experience.

Since things are targeted for Flow 5.0 (v18), we have a possibility to consider breaking/behavior altering changes. I would thus suggest that we'll override `setVisible(boolean visible)` for `PolymerTemplate` & `LitTemplate` and make it apply the CSS needed (either 1 or 2 in the earlier suggestions). This could be mentioned specially in the documentation. I don't even consider this to be a serious change and would IMO fine to backport it to 14 too since I agree with the comment from @Artur- about
>Being able to override server side visibility with CSS sounds more like a bug than a feature to me.

**Comment by pleku** [comment]
So we agreed on going with 1) with both keeping `hidden` and additionally setting style `display: none;`.

Once this lands to 18, the designer tutorial needs to be updated.
